test/contracts/kill_switch/kill_switch.test.js

```diff
@@ -0,0 +1,963 @@
+const { getNewProxyAddress, getEventArgument } = require('../../helpers/events')
+
+const ACL = artifacts.require('ACL')
+const Kernel = artifacts.require('Kernel')
+const DAOFactory = artifacts.require('DAOFactory')
+const KillSwitch = artifacts.require('KillSwitch')
+const IssuesRegistry = artifacts.require('IssuesRegistry')
+const KillSwitchedApp = artifacts.require('KillSwitchedAppMock')
+const EVMScriptRegistryFactory = artifacts.require('EVMScriptRegistryFactory')
+
+const FailingKillSwitchMock = artifacts.require('FailingKillSwitchMock')
+const KernelWithoutKillSwitchMock = artifacts.require('KernelWithoutKillSwitchMock')
+const KernelWithNonCompliantKillSwitchMock = artifacts.require('KernelWithNonCompliantKillSwitchMock')
+
+const ZERO_ADDRESS = '0x0000000000000000000000000000000000000000'
+
+contract('KillSwitch', ([_, root, owner, securityPartner, anyone]) => {
+  let dao, acl, app, registryFactory
+  let kernelBase, aclBase, appBase, killSwitchBase, issuesRegistryBase, daoFactory
+  let kernelWithoutKillSwitchBase, kernelWithNonCompliantKillSwitchBase, failingKillSwitchBase
+  let CORE_NAMESPACE, KERNEL_APP_ID, APP_MANAGER_ROLE, SET_SEVERITY_ROLE, SET_DEFAULT_ISSUES_REGISTRY_ROLE, SET_ISSUES_REGISTRY_ROLE, SET_ALLOWED_INSTANCES_ROLE, SET_DENIED_BASE_IMPLS_ROLE, SET_HIGHEST_ALLOWED_SEVERITY_ROLE
+
+  before('deploy base implementations', async () => {
+    // real
+    kernelBase = await Kernel.new(true) // petrify immediately
+    aclBase = await ACL.new()
+    registryFactory = await EVMScriptRegistryFactory.new()
+    killSwitchBase = await KillSwitch.new()
+    issuesRegistryBase = await IssuesRegistry.new()
+
+    // mocks
+    appBase = await KillSwitchedApp.new()
+    failingKillSwitchBase = await FailingKillSwitchMock.new()
+    kernelWithoutKillSwitchBase = await KernelWithoutKillSwitchMock.new()
+    kernelWithNonCompliantKillSwitchBase = await KernelWithNonCompliantKillSwitchMock.new()
+  })
+
+  before('load constants and roles', async () => {
+    CORE_NAMESPACE = await kernelBase.CORE_NAMESPACE()
+    KERNEL_APP_ID = await kernelBase.KERNEL_APP_ID()
+    APP_MANAGER_ROLE = await kernelBase.APP_MANAGER_ROLE()
+    SET_SEVERITY_ROLE = await issuesRegistryBase.SET_SEVERITY_ROLE()
+    SET_DEFAULT_ISSUES_REGISTRY_ROLE = await killSwitchBase.SET_DEFAULT_ISSUES_REGISTRY_ROLE()
+    SET_ISSUES_REGISTRY_ROLE = await killSwitchBase.SET_ISSUES_REGISTRY_ROLE()
+    SET_ALLOWED_INSTANCES_ROLE = await killSwitchBase.SET_ALLOWED_INSTANCES_ROLE()
+    SET_DENIED_BASE_IMPLS_ROLE = await killSwitchBase.SET_DENIED_BASE_IMPLS_ROLE()
+    SET_HIGHEST_ALLOWED_SEVERITY_ROLE = await killSwitchBase.SET_HIGHEST_ALLOWED_SEVERITY_ROLE()
+  })
+
+  context('when the kernel version does not support kill-switch logic', async () => {
+    before('create DAO factory', async () => {
+      daoFactory = await DAOFactory.new(kernelWithoutKillSwitchBase.address, aclBase.address, ZERO_ADDRESS, registryFactory.address)
```

Comment:
What do you think about no using the `DAOFactory` in these tests, but installing a `KillSwitch` manually, so that it's really only unit testing the `KillSwitch` (then the `DAOFactory` tests can act as more "integration" type tests).

Reply:
We could but I think we are covering more contracts and avoiding code duplication if we use it. However, if you prefer to follow that pattern I can refactor them.

Reply:
I'm wondering if it might be worth splitting the tests into `kill-switch/kill_switch.js`, which are mostly unit tests for the exposed `KillSwitch` functionality, and `kernel/kernel_kill_switch.js`, which are then the full suite of integration tests that interact with the various `Kernel`s / mocks.